components/merkledb/src/db.rs

```diff
@@ -470,6 +470,63 @@ enum NextIterValue {
 /// db.merge(fork.into_patch()).unwrap();
 /// ```
 ///
+/// # Merge Workflow
+///
+/// The user of a `Database` is responsible to ensure that forks are either created and merged
+/// sequentially or do not contain overlapping changes. By sequential creation we mean the following
+/// workflow:
+///
+/// ```
+/// # use exonum_merkledb::{Database, TemporaryDB};
+/// let db = TemporaryDB::new();
+/// let first_fork = db.fork();
+/// // Perform some operations on `first_fork`...
+/// db.merge(first_fork.into_patch()).unwrap();
+/// let second_fork = db.fork();
+/// // Perform some operations on `second_fork`...
+/// db.merge(second_fork.into_patch()).unwrap();
+/// ```
+///
+/// In contrast, this is a non-sequential workflow:
+///
+/// ```
+/// # use exonum_merkledb::{Database, TemporaryDB};
+/// let db = TemporaryDB::new();
+/// let first_fork = db.fork();
+/// // Perform some operations on `first_fork`...
+/// let second_fork = db.fork();
+/// // Perform some operations on `second_fork`...
+/// db.merge(first_fork.into_patch()).unwrap();
+/// db.merge(second_fork.into_patch()).unwrap();
+/// ```
+///
+/// In a non-sequential workflow, `first_fork` and `second_fork` **must not** contain overlapping
+/// changes (i.e., changes to the same index). If they do, the result of the merge may be
+/// unpredictable to the programmer and may break database invariants, e.g., that the length
+/// of an index is equal to the number of elements obtained by iterating over the index:
+///
+/// ```
+/// // NEVER USE THIS PATTERN!
+/// # use exonum_merkledb::{access::CopyAccessExt, Database, TemporaryDB};
+/// let db = TemporaryDB::new();
+/// let first_fork = db.fork();
+/// first_fork.get_list("list").extend(vec![1, 2, 3]);
+/// let second_fork = db.fork();
+/// second_fork.get_list("list").push(4);
+/// db.merge(first_fork.into_patch()).unwrap();
+/// db.merge(second_fork.into_patch()).unwrap();
+///
+/// let snapshot = db.snapshot();
+/// let list = snapshot.get_list::<_, i32>("list");
+/// assert_eq!(list.len(), 1);
```

Comment:
```suggestion
/// assert_eq!(list.len(), 3);
```
Shouldn't be 3 here ???

Reply:
Nope, that's the point of the code snippet: the concurrent merge workflow leads to logical bugs. That's why the top of the snippet contains a SHOUTING WARNING 🙃